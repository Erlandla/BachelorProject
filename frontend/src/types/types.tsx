export enum Status {
  newProblem = 'Nytt problem',
  started = 'Påbegynnt',
  solved = 'Løst',
}

export interface User {
  email: string
  telephone: string
  affiliation: string
}

export interface ContextUser {
  email: string
  password: string
  affiliation: string
  telephone: string
  isLoggedIn: boolean
  isAdmin: boolean
}

export interface ProblemCardProps {
  id: string
  title: string
  vendor: string
  status: Status
  specificProblem: string
  clearDataProduct: string
  accessibleData: string
  definedAction: string
  subCount: number
  owner: User
  subs: User[]
  edit: boolean
  approved: boolean
}

export interface IfetchType {
  limit: number
  searchPhrase: string
  categoryFilter: string
  orderBy: string
  email?: string
  relation?: number
  approved?: boolean
  similarProblem?: string
  filter?: number
}

export interface IfetchTypeMyProblems {
  limit: number
  searchPhrase: string
  categoryFilter: string
  orderBy: string
  email?: string
  relation?: number
}

export interface Categories {
  specificProblem: string[]
  accessibleData: string[]
  dataProduct: string[]
}

export interface AdminInfo {
  title: string
  count: number
  link: string
}
