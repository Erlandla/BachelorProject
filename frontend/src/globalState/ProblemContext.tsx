import React from 'react'
import { createContext, type Dispatch, type SetStateAction, type ReactNode, useState } from 'react'
import { type ProblemCardProps, Status, type User, type ContextUser } from '../types/types'

export interface ProblemContextInterface {
  problem: ProblemCardProps
  setProblem: Dispatch<SetStateAction<ProblemCardProps>>

  user: ContextUser
  setUser: Dispatch<SetStateAction<ContextUser>>
}

const initialUser: User = {
  email: '-',
  telephone: '-',
  affiliation: '-',
}

const initialSubs = [initialUser]

const initialProblemCard: ProblemContextInterface = {
  problem: {
    id: '',
    title: '',
    system: '',
    status: Status.newProblem,
    specificProblem: '',
    clearDataProduct: '',
    accessibleData: '',
    definedAction: '',
    subCount: 0,
    owner: initialUser,
    subs: initialSubs,
  },
  setProblem: (_problem: ProblemCardProps) => {},

  user: {
    email: '-',
    password: '-',
    affiliation: '-',
    telephone: '-',
    isLoggedIn: false,
    isAdmin: false,
  },
  setUser: (_user: ContextUser) => {},
} as any // as ProblemContextInferface
// Which had an eslint error Always prefer const x: T = { ... }. the assertion to any is ignored by objectLiteralTypeAssertions.
// https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-assertions.md

export const ProblemContext = createContext(initialProblemCard)

interface ProblemProviderProps {
  children: ReactNode
}

export default function ProblemProvider({ children }: ProblemProviderProps) {
  const [user, setUser] = useState({
    email: '',
    password: '',
    affiliation: '',
    telephone: '',
    isLoggedIn: false,
    isAdmin: false,
  })

  const [problem, setProblem] = useState<ProblemCardProps>({
    id: '',
    title: 'Tittel',
    vendor: '',
    status: Status.newProblem,
    specificProblem: '',
    clearDataProduct: '',
    accessibleData: '',
    definedAction: '',
    subCount: 0,
    owner: initialUser,
    subs: initialSubs,
    edit: false,
    approved: false,
  })

  return (
    <ProblemContext.Provider value={{ problem, setProblem, user, setUser }}>
      {children}
    </ProblemContext.Provider>
  )
}
