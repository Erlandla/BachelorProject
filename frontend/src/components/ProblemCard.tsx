import React, { useEffect, useState, useContext } from 'react'
import PersonalVideoIcon from '@mui/icons-material/PersonalVideo'
import PeopleIcon from '@mui/icons-material/People'
import LocationCityIcon from '@mui/icons-material/LocationCity'
import { type ProblemCardProps } from '../types/types'
import { Status } from '../types/types'
import ODACircle from './ODACircle'
import { ProblemContext } from '../globalState/ProblemContext'
import { Link } from 'react-router-dom'
import validateProblem from '../utils/validateProblem'

function ProblemCard(props: ProblemCardProps) {
  const [statusColor, setStatusColor] = useState(
    'rounded-full flex items-center justify-center h-4 w-4 mr-2 bg-statusRed'
  )

  const { setProblem } = useContext(ProblemContext)

  const onChange = () => {
    setProblem({
      id: props.id,
      title: props.title,
      vendor: props.vendor,
      status: props.status,
      specificProblem: props.specificProblem,
      clearDataProduct: props.clearDataProduct,
      accessibleData: props.accessibleData,
      definedAction: props.definedAction,
      subCount: props.subCount,
      owner: props.owner,
      subs: props.subs,
      edit: props.edit,
      approved: props.approved,
    })
    document.body.scrollTop = 0 // For Safari
    document.documentElement.scrollTop = 0 // for safari, chrome, edge, etc.
  }

  useEffect(() => {
    if (props.approved.toString() === 'true') {
      switch (props.status) {
        case Status.newProblem:
          setStatusColor('rounded-full flex items-center justify-center h-4 w-4 mr-2 bg-statusRed')
          break

        case Status.started:
          setStatusColor(
            'rounded-full flex items-center justify-center h-4 w-4 mr-2 bg-statusOrange'
          )
          break

        case Status.solved:
          setStatusColor(
            'rounded-full flex items-center justify-center h-4 w-4 mr-2 bg-statusGreen'
          )
          break
      }
    } else {
      setStatusColor('rounded-full flex items-center justify-center h-4 w-4 mr-2 bg-white')
    }
  }, [props.status])

  try {
    validateProblem(props)
  } catch (error: any) {
    const errorMsg = 'En feil oppstod: '.concat(error)
    return (
      <div>
        <p className="color-red items-center">{errorMsg}</p>
      </div>
    )
  }

  return (
    <div className="flex flex-col justify-between h-64 w-80 bg-buttonDark px-5 py-5 text-white cursor-pointer hover:drop-shadow-3xl hover:bg-buttonHover">
      <Link to={props.edit ? '/GodkjennProblem' : '/InspiserProblem'} onClick={onChange}>
        <section className="flex justify-between  items-center">
          <p className="font-bold text-left truncate max-w-[15ch]">{props.title}</p>

          <div className="flex flex-row items-center">
            <ODACircle style={statusColor} text={''} />
            <p className="whitespace-nowrap">
              {props.approved.toString() === 'true' ? props.status : 'Til vurdering'}
            </p>
          </div>
        </section>

        <section className="text-left pt-2.5 h-44">
          <p className="text-sm line-clamp-6">{props.specificProblem}</p>
        </section>

        <section className=" items-end flex flex-row justify-between text-xs">
          <div className="flex flex-row">
            <PersonalVideoIcon sx={{ fontSize: '1rem' }} />
            <p className="ml-1.5 whitespace-nowrap">{props.vendor}</p>
          </div>
          <p>&#x2022;</p>
          <div className="flex flex-row">
            <PeopleIcon sx={{ fontSize: '1rem' }} />
            <p className="ml-1.5 whitespace-nowrap">{props.subCount}</p>
          </div>
          <p>&#x2022;</p>
          <div className="flex flex-row">
            <LocationCityIcon sx={{ fontSize: '1rem' }} />
            <p className="ml-1.5 whitespace-nowrap truncate max-w-[20ch]">
              {props.owner.affiliation}
            </p>
          </div>
        </section>
      </Link>
    </div>
  )
}

export default ProblemCard
