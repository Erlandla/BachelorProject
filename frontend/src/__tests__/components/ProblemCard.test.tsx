import React from 'react'
import { cleanup, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ProblemCard from '../../components/ProblemCard'
import { ContextUser, Status, type ProblemCardProps } from '../../types/types'
import { MemoryRouter } from 'react-router-dom'
import { ProblemContext } from '../../globalState/ProblemContext'
import renderer from 'react-test-renderer'

// ChallengeCardProp-objects used for various test cases
const problemCorrectData: ProblemCardProps = {
  id: '123',
  title: 'Test Challenge',
  vendor: 'Test System',
  status: Status.newProblem,
  specificProblem: 'Test Specific Problem',
  clearDataProduct: 'Test Clear Data Product',
  accessibleData: 'Test Accessible Data',
  definedAction: 'Test Defined Action',
  subCount: 5,
  owner: { email: 'test@mail.no', telephone: '111 22 333', affiliation: 'Test kommune' },
  subs: [],
  edit: false,
  approved: true
}
const problemEmptyTitle: ProblemCardProps = {
  id: '1234',
  title: '',
  vendor: 'aa',
  status: Status.newProblem,
  specificProblem: 'aa',
  clearDataProduct: 'aa',
  accessibleData: 'aa',
  definedAction: 'aa',
  subCount: 5,
  owner: { email: 'test@mail.no', telephone: '111 22 333', affiliation: 'aaaaaa' },
  subs: [],
  edit: false,
  approved: false
}
const problemNegativeSubCount: ProblemCardProps = {
  id: '12345',
  title: 'Test Challenge',
  vendor: 'Test System',
  status: Status.newProblem,
  specificProblem: 'Test Specific Problem',
  clearDataProduct: 'Test Clear Data Product',
  accessibleData: 'Test Accessible Data',
  definedAction: 'Test Defined Action',
  subCount: -5,
  owner: { email: 'test@mail.no', telephone: '111 22 333', affiliation: 'Test kommune' },
  subs: [],
  edit: false,
  approved: false
}
const problemInvalidEmail: ProblemCardProps = {
  id: '12345',
  title: 'Test Challenge',
  vendor: 'Test System',
  status: Status.newProblem,
  specificProblem: 'Test Specific Problem',
  clearDataProduct: 'Test Clear Data Product',
  accessibleData: 'Test Accessible Data',
  definedAction: 'Test Defined Action',
  subCount: 5,
  owner: { email: 'c:', telephone: '111 22 333', affiliation: 'Test kommune' },
  subs: [],
  edit: false,
  approved: false
}
const problemInvalidPhoneNumb: ProblemCardProps = {
  id: '12345',
  title: 'Test Challenge',
  vendor: 'Test System',
  status: Status.newProblem,
  specificProblem: 'Test Specific Problem',
  clearDataProduct: 'Test Clear Data Product',
  accessibleData: 'Test Accessible Data',
  definedAction: 'Test Defined Action',
  subCount: 5,
  owner: { email: 'test@ail.no', telephone: '11333', affiliation: 'Test kommune' },
  subs: [],
  edit: false,
  approved: false
}
const loggedInUser: ContextUser = {
  email: 'a@b.c',
  password: 'aaaaaaaaaaaaaaa',
  affiliation: 'TRONDHJÃ†M, BAYBEY',
  telephone: '111 22 333',
  isLoggedIn: true,
  isAdmin: true,
}

afterEach(() => {
  cleanup()
})

describe('Renders with no errors', () => {
  it('renders challenge card with correct data', () => {
    render(
      <MemoryRouter>
        <ProblemCard {...problemCorrectData} />
      </MemoryRouter>
    )
    expect(screen.getByText(problemCorrectData.title)).toBeInTheDocument()
    expect(screen.getByText(problemCorrectData.status)).toBeInTheDocument()
    expect(screen.getByText(problemCorrectData.specificProblem)).toBeInTheDocument()
    expect(screen.getByText(problemCorrectData.vendor)).toBeInTheDocument()
    expect(screen.getByText(problemCorrectData.subCount)).toBeInTheDocument()
    expect(screen.getByText(problemCorrectData.owner.affiliation)).toBeInTheDocument()
  })

  it('calls setChallenge on click', () => {
    const setProblemMock = jest.fn()
    render(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemCorrectData,
            setProblem: setProblemMock,
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemCorrectData} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    userEvent.click(screen.getByText(problemCorrectData.title))
    expect(setProblemMock).toHaveBeenCalledWith(problemCorrectData)
  })
})

describe('Weird cases of data in ProblemCardProps', () => {
  it('Title-field is an empty string', () => {
    render(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemEmptyTitle,
            setProblem: jest.fn(),
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemEmptyTitle} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    expect(screen.getByText('En feil oppstod: Error: Tittel-feltet er tomt'))
  })

  it('Subs-count is negative', () => {
    render(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemNegativeSubCount,
            setProblem: jest.fn(),
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemNegativeSubCount} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    expect(screen.getByText('En feil oppstod: Error: Mengden abonnenter er negativ'))
  })

  it('Owner with invalid email', () => {
    render(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemInvalidEmail,
            setProblem: jest.fn(),
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemInvalidEmail} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    expect(screen.getByText('En feil oppstod: Error: Ugyldig email'))
  })

  it('Owner with invalid phone number', () => {
    render(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemInvalidPhoneNumb,
            setProblem: jest.fn(),
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemInvalidPhoneNumb} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    expect(screen.getByText('En feil oppstod: Error: Ugyldig telefonnummer'))
  })
})

it('Matches snapshot', () => {
  const snapshot = renderer
    .create(
      <MemoryRouter>
        <ProblemContext.Provider
          value={{
            problem: problemInvalidPhoneNumb,
            setProblem: jest.fn(),
            user: loggedInUser,
            setUser: jest.fn(),
          }}
        >
          <ProblemCard {...problemInvalidPhoneNumb} />
        </ProblemContext.Provider>
      </MemoryRouter>
    )
    .toJSON()
  expect(snapshot).toMatchSnapshot()
})
