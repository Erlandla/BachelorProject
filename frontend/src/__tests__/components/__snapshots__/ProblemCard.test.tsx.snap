// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div>
  <p
    className="color-red items-center"
  >
    En feil oppstod: Error: Ugyldig telefonnummer
  </p>
</div>
`;
